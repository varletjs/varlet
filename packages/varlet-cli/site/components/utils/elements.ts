import { isNumber, isString, toNumber, kebabCase } from './shared'
import type { StyleVars } from '../../utils'

export function getLeft(element: HTMLElement): number {
  const { left } = element.getBoundingClientRect()

  return left + (document.body.scrollLeft || document.documentElement.scrollLeft)
}

export const isRem = (value: unknown) => isString(value) && value.endsWith('rem')

// example 1 || 1px
export const isPx = (value: unknown) => (isString(value) && value.endsWith('px')) || isNumber(value)

// example 1%
export const isPercent = (value: unknown) => isString(value) && value.endsWith('%')

// example 1vw
export const isVw = (value: unknown) => isString(value) && value.endsWith('vw')

// example 1vh
export const isVh = (value: unknown) => isString(value) && value.endsWith('vh')

// example return 1
export const toPxNum = (value: unknown): number => {
  if (isNumber(value)) {
    return value
  }

  if (isPx(value)) {
    return +(value as string).replace('px', '')
  }

  if (isVw(value)) {
    return (+(value as string).replace('vw', '') * window.innerWidth) / 100
  }

  if (isVh(value)) {
    return (+(value as string).replace('vh', '') * window.innerHeight) / 100
  }

  if (isRem(value)) {
    const num = +(value as string).replace('rem', '')
    const rootFontSize = window.getComputedStyle(document.documentElement).fontSize

    return num * parseFloat(rootFontSize)
  }

  if (isString(value)) {
    return toNumber(value)
  }

  // % and other
  return 0
}

// example return 1px 1% 1vw 1vh 1rem null
export const toSizeUnit = (value: unknown) => {
  if (value == null) {
    return null
  }

  if (isPercent(value) || isVw(value) || isVh(value) || isRem(value)) {
    return value
  }

  return `${toPxNum(value)}px`
}

export function formatStyleVars(styleVars: StyleVars | null) {
  return Object.entries(styleVars ?? {}).reduce((styles, [key, value]) => {
    const cssVar = key.startsWith('--') ? key : `--${kebabCase(key)}`
    styles[cssVar] = value

    return styles
  }, {} as StyleVars)
}

export function doubleRaf() {
  return new Promise((resolve) => {
    requestAnimationFrame(() => {
      requestAnimationFrame(resolve)
    })
  })
}
