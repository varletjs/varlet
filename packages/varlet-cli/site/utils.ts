import { get } from 'lodash-es'
import { StyleProvider, type Theme, withSiteConfigNamespace } from '@varlet/cli/client'

export interface Menu {
  doc: string
  text: Record<string, string>
  type: MenuTypes
  useMobile?: boolean
}

export enum MenuTypes {
  TITLE = 1,
  COMPONENT = 2,
  DOCUMENTATION = 3,
}

export function removeEmpty(object: Record<string, string> = {}) {
  return Object.keys(object).reduce((record: Record<string, string>, key) => {
    const value = object[key]
    value && (record[key] = value)
    return record
  }, {})
}

export function isPhone() {
  return /Android|webOS|iPhone|iPod|BlackBerry|Pad/i.test(navigator.userAgent)
}

export function inIframe() {
  return window.self !== window.top
}

export function setTheme(config: Record<string, any>, name: Theme) {
  const styleVars = withSiteConfigNamespace(get(config, name, {}))

  StyleProvider(styleVars)
}

export function utoa(data: string): string {
  return btoa(unescape(encodeURIComponent(data)))
}
