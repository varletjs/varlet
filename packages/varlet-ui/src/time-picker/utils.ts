import { hours24, hoursAmpm } from './props'
import type { AmPm, AllowedTime } from './props'
import { toNumber } from '@varlet/shared'

type DisableProps = {
  time: number
  format: string
  ampm: AmPm | undefined
  hour: string
  minute: number
  max: string | undefined
  min: string | undefined
  disableHour: Array<string>
  allowedTime: AllowedTime | undefined
}

export const notConvert = (format: string, ampm: AmPm | undefined): boolean => format === '24hr' || ampm === 'am'

export const convertHour = (format: string, ampm: AmPm | undefined, hour: string) => {
  const index = hoursAmpm.findIndex((hourAmpm) => toNumber(hourAmpm) === toNumber(hour))
  const getHour = notConvert(format, ampm) ? hour : hours24[index]

  return {
    hourStr: getHour,
    hourNum: toNumber(getHour),
  }
}

export const getNumberTime = (time: string) => {
  const [hour, minute, second] = time.split(':')

  return {
    hour: toNumber(hour),
    minute: toNumber(minute),
    second: toNumber(second),
  }
}

export const getIsDisableMinute = (values: Omit<DisableProps, 'minute'>): boolean => {
  const { time, format, ampm, hour, max, min, disableHour } = values
  const { hourStr, hourNum } = convertHour(format, ampm, hour)
  let isBetweenMinMax = false
  let isAllow = false

  if (disableHour.includes(hourStr)) return true

  if (max && !min) {
    const { hour: maxHour, minute: maxMinute } = getNumberTime(max)
    isBetweenMinMax = maxHour === hourNum && time > maxMinute
  }

  if (!max && min) {
    const { hour: minHour, minute: minMinute } = getNumberTime(min)

    isBetweenMinMax = minHour === hourNum && time < minMinute
  }

  if (max && min) {
    const { hour: maxHour, minute: maxMinute } = getNumberTime(max)
    const { hour: minHour, minute: minMinute } = getNumberTime(min)

    isBetweenMinMax = (minHour === hourNum && time < minMinute) || (maxHour === hourNum && time > maxMinute)
  }

  if (values.allowedTime?.minutes) isAllow = values.allowedTime?.minutes(time)

  return isBetweenMinMax || isAllow
}

export const getIsDisableSecond = (values: DisableProps): boolean => {
  const { time, format, ampm, hour, minute, max, min, disableHour } = values
  const { hourStr, hourNum } = convertHour(format, ampm, hour)
  let isBetweenMinMax = false
  let isAllow = false

  if (disableHour.includes(hourStr)) return true

  if (max && !min) {
    const { hour: maxHour, minute: maxMinute, second: maxSecond } = getNumberTime(max)

    isBetweenMinMax = (maxHour === hourNum && maxMinute < minute) || (maxMinute === minute && time > maxSecond)
  }

  if (!max && min) {
    const { hour: minHour, minute: minMinute, second: minSecond } = getNumberTime(min)

    isBetweenMinMax = (minHour === hourNum && minMinute > minute) || (minMinute === minute && time > minSecond)
  }

  if (max && min) {
    const { hour: maxHour, minute: maxMinute, second: maxSecond } = getNumberTime(max)
    const { hour: minHour, minute: minMinute, second: minSecond } = getNumberTime(min)

    isBetweenMinMax =
      (maxHour === hourNum && maxMinute < minute) ||
      (minHour === hourNum && minMinute > minute) ||
      (maxHour === hourNum && maxMinute === minute && time > maxSecond) ||
      (minHour === hourNum && minMinute === minute && time < minSecond)
  }

  if (values.allowedTime?.seconds) isAllow = values.allowedTime?.seconds(time)

  return isBetweenMinMax || isAllow
}
