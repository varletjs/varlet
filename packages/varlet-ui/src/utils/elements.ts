import { isNumber, isObject, isString, kebabCase, toNumber } from '@varlet/shared'
import { getGlobalThis } from './shared'
import type { StyleVars } from '../style-provider'

export function getLeft(element: HTMLElement): number {
  const { left } = element.getBoundingClientRect()

  return left + (document.body.scrollLeft || document.documentElement.scrollLeft)
}

export function getTop(element: HTMLElement): number {
  const { top } = element.getBoundingClientRect()

  return top + (document.body.scrollTop || document.documentElement.scrollTop)
}

export function getScrollTop(element: Element | Window): number {
  const top = 'scrollTop' in element ? element.scrollTop : element.pageYOffset

  // iOS scroll bounce cause minus scrollTop
  return Math.max(top, 0)
}

export function getScrollLeft(element: Element | Window): number {
  const left = 'scrollLeft' in element ? element.scrollLeft : element.pageXOffset

  return Math.max(left, 0)
}

export async function inViewport(element: HTMLElement): Promise<boolean> {
  await doubleRaf()
  const { top, bottom, left, right } = element.getBoundingClientRect()
  const { innerWidth, innerHeight } = window

  const xInViewport = left <= innerWidth && right >= 0
  const yInViewport = top <= innerHeight && bottom >= 0

  return xInViewport && yInViewport
}

export function getTranslate(el: HTMLElement) {
  const { transform } = window.getComputedStyle(el)
  return +transform.slice(transform.lastIndexOf(',') + 2, transform.length - 1)
}

export function getParentScroller(el: HTMLElement): HTMLElement | Window {
  let element = el

  while (element) {
    if (!element.parentNode) {
      break
    }

    element = element.parentNode as HTMLElement

    if (element === document.body || element === document.documentElement) {
      break
    }

    const scrollRE = /(scroll|auto)/
    const { overflowY, overflow } = window.getComputedStyle(element)

    if (scrollRE.test(overflowY) || scrollRE.test(overflow)) {
      return element
    }
  }

  return window
}

export function getAllParentScroller(el: HTMLElement): Array<HTMLElement | Window> {
  const allParentScroller: Array<HTMLElement | Window> = []
  let element: HTMLElement | Window = el

  while (element !== window) {
    element = getParentScroller(element as HTMLElement)
    allParentScroller.push(element)
  }

  return allParentScroller
}

export function getTarget(target: string | HTMLElement, componentName: string) {
  if (isString(target)) {
    const el = document.querySelector(target)

    if (!el) {
      throw Error(`[Varlet] ${componentName}: target element cannot found`)
    }

    return el as HTMLElement
  }

  if (isObject(target)) return target

  throw Error(`[Varlet] ${componentName}: type of prop "target" should be a selector or an element object`)
}

// example 1rem
export const isRem = (value: unknown): value is string => isString(value) && value.endsWith('rem')

// e.g. 1 || 1px
export const isPx = (value: unknown): value is string | number =>
  (isString(value) && value.endsWith('px')) || isNumber(value)

// e.g. 1%
export const isPercent = (value: unknown): value is string => isString(value) && value.endsWith('%')

// e.g. 1vw
export const isVw = (value: unknown): value is string => isString(value) && value.endsWith('vw')

// e.g. 1vh
export const isVh = (value: unknown): value is string => isString(value) && value.endsWith('vh')

// e.g. calc(1px + 1px)
export const isCalc = (value: unknown): value is string => isString(value) && value.startsWith('calc(')

// e.g. var(--color-primary)
export const isVar = (value: unknown): value is string => isString(value) && value.startsWith('var(')

// e.g. return 1
export const toPxNum = (value: unknown): number => {
  if (isNumber(value)) {
    return value
  }

  if (isPx(value)) {
    return +(value as string).replace('px', '')
  }

  if (isVw(value)) {
    return (+(value as string).replace('vw', '') * window.innerWidth) / 100
  }

  if (isVh(value)) {
    return (+(value as string).replace('vh', '') * window.innerHeight) / 100
  }

  if (isRem(value)) {
    const num = +(value as string).replace('rem', '')
    const rootFontSize = window.getComputedStyle(document.documentElement).fontSize

    return num * parseFloat(rootFontSize)
  }

  if (isString(value)) {
    return toNumber(value)
  }

  // % and other
  return 0
}

// e.g. return 1px 1% 1vw 1vh 1rem null var(--color-primary) calc(1px + 1px)
export const toSizeUnit = (value: unknown) => {
  if (value == null) {
    return undefined
  }

  if (isPercent(value) || isVw(value) || isVh(value) || isRem(value) || isCalc(value) || isVar(value)) {
    return value
  }

  return `${toPxNum(value)}px`
}

export const multiplySizeUnit = (value: unknown, quantity = 1) => {
  if (value == null) {
    return undefined
  }

  const legalSize = toSizeUnit(value) as string

  const unit = legalSize.match(/(vh|%|rem|px|vw)$/)![0]

  return `${parseFloat(legalSize) * quantity}${unit}`
}

export function requestAnimationFrame(fn: FrameRequestCallback): number {
  const globalThis = getGlobalThis()

  return globalThis.requestAnimationFrame ? globalThis.requestAnimationFrame(fn) : globalThis.setTimeout(fn, 16)
}

export function cancelAnimationFrame(handle: number): void {
  const globalThis = getGlobalThis()

  globalThis.cancelAnimationFrame ? globalThis.cancelAnimationFrame(handle) : globalThis.clearTimeout(handle)
}

export function nextTickFrame(fn: FrameRequestCallback) {
  requestAnimationFrame(() => {
    requestAnimationFrame(fn)
  })
}

export function doubleRaf() {
  return new Promise((resolve) => {
    requestAnimationFrame(() => {
      requestAnimationFrame(resolve)
    })
  })
}

export function raf() {
  return new Promise((resolve) => {
    requestAnimationFrame(resolve)
  })
}

interface ScrollToOptions {
  top?: number
  left?: number
  duration?: number
  animation: (progress: number) => number
}

export function scrollTo(
  element: HTMLElement | Window,
  { top = 0, left = 0, duration = 300, animation }: ScrollToOptions
): Promise<void> {
  const startTime = Date.now()

  const scrollTop = getScrollTop(element)
  const scrollLeft = getScrollLeft(element)

  return new Promise((resolve) => {
    const frame = () => {
      const progress = (Date.now() - startTime) / duration

      if (progress < 1) {
        const nextTop = scrollTop + (top - scrollTop) * animation(progress)
        const nextLeft = scrollLeft + (left - scrollLeft) * animation(progress)

        element.scrollTo(nextLeft, nextTop)
        requestAnimationFrame(frame)
      } else {
        element.scrollTo(left, top)
        resolve()
      }
    }

    requestAnimationFrame(frame)
  })
}

export function formatStyleVars(styleVars: StyleVars | null) {
  return Object.entries(styleVars ?? {}).reduce((styles, [key, value]) => {
    const cssVar = key.startsWith('--') ? key : `--${kebabCase(key)}`
    styles[cssVar] = value

    return styles
  }, {} as StyleVars)
}

export function supportTouch() {
  const inBrowser = typeof window !== 'undefined'
  return inBrowser && 'ontouchstart' in window
}

export function padStartFlex(style: string | undefined) {
  return style === 'start' || style === 'end' ? `flex-${style}` : style
}
