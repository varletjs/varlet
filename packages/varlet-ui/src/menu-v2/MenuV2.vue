<template>
  <div ref="host" :class="n()" @click="handleClick" @mouseenter="handleMouseenter" @mouseleave="handleMouseleave">
    <slot />

    <teleport :to="teleport">
      <transition :name="n()" @after-enter="onOpened" @after-leave="onClosed">
        <div
          ref="menu"
          :style="{ zIndex }"
          :class="classes(n('menu'), [defaultStyle, n('--menu-background-color')], 'var-elevation--3')"
          v-show="show"
          @click.stop
          @mouseenter="handleMenuMouseenter"
          @mouseleave="handleMenuMouseleave"
        >
          <slot name="menu" />
        </div>
      </transition>
    </teleport>
  </div>
</template>

<script lang="ts">
import flip from '@popperjs/core/lib/modifiers/flip'
import offset from '@popperjs/core/lib/modifiers/offset'
import { createPopper } from '@popperjs/core/lib/popper-lite'
import { call, createNamespace } from '../utils/components'
import { defineComponent, onMounted, onUnmounted, ref, watch } from 'vue'
import { props } from './props'
import { doubleRaf, toPxNum } from '../utils/elements'
import { useZIndex } from '../context/zIndex'
import type { Ref } from 'vue'
import type { Instance } from '@popperjs/core/lib/types'
import type { NeededPopperPlacement } from './props'

const { n, classes } = createNamespace('menu-v2')

export interface HostSize {
  width: number
  height: number
}

export interface Position {
  placement: NeededPopperPlacement
  skidding: number
  distance: number
}

export default defineComponent({
  name: 'VarMenuV2',
  props,
  setup(props) {
    const host: Ref<null | HTMLElement> = ref(null)
    const menu: Ref<null | HTMLElement> = ref(null)
    const show: Ref<boolean> = ref(false)
    const { zIndex } = useZIndex(() => show.value, 1)

    let popover: Instance | null = null
    let clickSelf = false
    let enterMenu = false
    let enterHost = false
    let hostSize: HostSize = { width: 0, height: 0 }

    const computeHostSize = () => {
      const { width, height } = getComputedStyle(host.value!)

      hostSize = {
        width: toPxNum(width),
        height: toPxNum(height),
      }
    }

    const handleMouseenter = () => {
      if (props.trigger !== 'hover') {
        return
      }

      enterHost = true

      open()
    }

    const handleMouseleave = async () => {
      if (props.trigger !== 'hover') {
        return
      }

      enterHost = false

      await doubleRaf()

      if (enterMenu) {
        return
      }

      close()
    }

    const handleMenuMouseenter = () => {
      if (props.trigger !== 'hover') {
        return
      }

      enterMenu = true
    }

    const handleMenuMouseleave = async () => {
      if (props.trigger !== 'hover') {
        return
      }

      enterMenu = false

      await doubleRaf()

      if (enterHost) {
        return
      }

      close()
    }

    const handleClick = () => {
      open()
      clickSelf = true
    }

    const handleMenuClose = () => {
      if (clickSelf) {
        clickSelf = false
        return
      }

      show.value = false
      call(props['onUpdate:show'], false)
    }

    const getPosition = (): Position => {
      computeHostSize()

      const offset = {
        x: toPxNum(props.offsetX),
        y: toPxNum(props.offsetY),
      }

      switch (props.placement) {
        case 'cover-top':
          return {
            placement: 'bottom',
            skidding: offset.x,
            distance: offset.y - hostSize.height,
          }

        case 'cover-top-start':
          return {
            placement: 'bottom-start',
            skidding: offset.x,
            distance: offset.y - hostSize.height,
          }

        case 'cover-top-end':
          return {
            placement: 'bottom-end',
            skidding: offset.x,
            distance: offset.y - hostSize.height,
          }

        case 'cover-bottom':
          return {
            placement: 'top',
            skidding: offset.x,
            distance: -offset.y - hostSize.height,
          }

        case 'cover-bottom-start':
          return {
            placement: 'top-start',
            skidding: offset.x,
            distance: -offset.y - hostSize.height,
          }

        case 'cover-bottom-end':
          return {
            placement: 'top-end',
            skidding: offset.x,
            distance: -offset.y - hostSize.height,
          }

        case 'cover-left':
          return {
            placement: 'right',
            skidding: offset.y,
            distance: offset.x - hostSize.width,
          }

        case 'cover-right':
          return {
            placement: 'left',
            skidding: offset.y,
            distance: -offset.x - hostSize.width,
          }

        case 'left':
        case 'left-start':
        case 'left-end':
          return {
            placement: props.placement,
            skidding: offset.y,
            distance: -offset.x,
          }

        case 'top':
        case 'top-start':
        case 'top-end':
          return {
            placement: props.placement,
            skidding: offset.x,
            distance: -offset.y,
          }

        case 'bottom':
        case 'bottom-start':
        case 'bottom-end':
          return {
            placement: props.placement,
            skidding: offset.x,
            distance: offset.y,
          }

        case 'right':
        case 'right-start':
        case 'right-end':
          return {
            placement: props.placement,
            skidding: offset.y,
            distance: offset.x,
          }
      }
    }

    const getPopperOptions = () => {
      const { placement, skidding, distance } = getPosition()

      return {
        placement,
        modifiers: [
          {
            ...flip,
            enabled: show.value,
          },
          {
            ...offset,
            options: {
              offset: [skidding, distance],
            },
          },
        ],
      }
    }

    // expose
    const resize = () => {
      popover!.setOptions(getPopperOptions())
    }

    // expose
    const open = () => {
      const { disabled } = props

      if (disabled) {
        return
      }

      show.value = true
      call(props['onUpdate:show'], true)
    }

    // expose
    const close = () => {
      show.value = false
      call(props['onUpdate:show'], false)
    }

    watch(
      () => props.show,
      (newValue) => {
        show.value = newValue ?? false
      },
      { immediate: true }
    )

    watch(() => props.offsetX, resize)
    watch(() => props.offsetY, resize)
    watch(() => props.placement, resize)

    watch(
      () => show.value,
      (newValue) => {
        if (newValue) {
          resize()
          call(props.onOpen)
        } else {
          call(props.onClose)
        }
      }
    )

    watch(
      () => props.trigger,
      (newValue) => {
        if (newValue === 'click') {
          document.addEventListener('click', handleMenuClose)
        } else {
          document.removeEventListener('click', handleMenuClose)
        }
      }
    )

    watch(() => props.disabled, close)

    onMounted(() => {
      popover = createPopper(host.value!, menu.value!, getPopperOptions())

      if (props.trigger === 'click') {
        document.addEventListener('click', handleMenuClose)
      }
    })
    onUnmounted(() => {
      document.removeEventListener('click', handleMenuClose)
      popover!.destroy()
    })

    return {
      menu,
      host,
      hostSize,
      show,
      zIndex,
      n,
      classes,
      handleClick,
      handleMenuClose,
      handleMouseenter,
      handleMouseleave,
      handleMenuMouseenter,
      handleMenuMouseleave,
      resize,
      open,
      close,
    }
  },
})
</script>

<style lang="less">
@import './menuV2';
@import '../styles/elevation.less';
@import '../styles/common.less';
</style>
