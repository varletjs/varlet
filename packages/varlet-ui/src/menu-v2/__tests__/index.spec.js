import MenuV2 from '..'
import { createApp, h } from 'vue'
import { mount } from '@vue/test-utils'
import { delay, mockStubs, trigger } from '../../utils/jest'
import { doubleRaf } from '../../utils/elements'

test('test menu v2 plugin', () => {
  const app = createApp({}).use(MenuV2)
  expect(app.component(MenuV2.name)).toBeTruthy()
})

test('test menu v2 placement', async () => {
  const { mockRestore } = mockStubs()

  for (const placement of [
    'top',
    'top-start',
    'top-end',
    'bottom',
    'bottom-start',
    'bottom-end',
    'right',
    'right-start',
    'right-end',
    'left',
    'left-start',
    'left-end',
    'cover-top',
    'cover-top-start',
    'cover-top-end',
    'cover-bottom',
    'cover-bottom-start',
    'cover-bottom-end',
    'cover-left',
    'cover-right',
  ]) {
    const root = document.createElement('div')

    mount(MenuV2, {
      props: {
        placement,
        teleport: root,
      },
    })

    await doubleRaf()

    expect(root.innerHTML).toMatchSnapshot()
  }

  mockRestore()
})

test('test menu v2 click trigger', async () => {
  const { mockRestore } = mockStubs()

  const wrapper = mount(MenuV2, {
    attachTo: document.body,
  })

  await wrapper.trigger('click')
  await delay(300)
  expect(document.body.innerHTML).toMatchSnapshot()

  await trigger(document, 'click')
  await delay(300)
  expect(document.body.innerHTML).toMatchSnapshot()

  document.body.innerHTML = ''

  mockRestore()
})

test('test menu v2 hover trigger and events', async () => {
  const { mockRestore } = mockStubs()

  const onOpen = jest.fn()
  const onOpened = jest.fn()
  const onClose = jest.fn()
  const onClosed = jest.fn()

  const root = document.createElement('div')

  const wrapper = mount(MenuV2, {
    props: {
      trigger: 'hover',
      teleport: root,
      onOpen,
      onOpened,
      onClose,
      onClosed,
    },
  })

  await wrapper.trigger('mouseenter')
  expect(onOpen).toHaveBeenCalledTimes(1)

  await delay(300)

  expect(onOpened).toHaveBeenCalledTimes(1)
  expect(root.innerHTML).toMatchSnapshot()

  await wrapper.trigger('mouseleave')

  await doubleRaf()
  await delay(0)
  expect(onClose).toHaveBeenCalledTimes(1)
  await delay(300)

  expect(onClosed).toHaveBeenCalledTimes(1)
  expect(root.innerHTML).toMatchSnapshot()

  wrapper.unmount()
  mockRestore()
})

test('test menu v2 default style', async () => {
  const root = document.createElement('div')

  mount(MenuV2, {
    props: {
      defaultStyle: false,
      teleport: root,
    },
  })

  await doubleRaf()

  expect(root.innerHTML).toMatchSnapshot()
})

test('test menu v2 offset', async () => {
  const { mockRestore } = mockStubs()

  const root = document.createElement('div')

  mount(MenuV2, {
    props: {
      offsetX: 100,
      offsetY: 100,
      teleport: root,
    },
  })

  await doubleRaf()

  expect(root.innerHTML).toMatchSnapshot()

  mockRestore()
})

test('test menu v2 hover the menu list', async () => {
  const { mockRestore } = mockStubs()

  const root = document.createElement('div')

  mount(MenuV2, {
    props: {
      show: true,
      trigger: 'hover',
      teleport: root,
    },
  })

  await doubleRaf()
  await trigger(root.querySelector('.var-menu-v2__menu'), 'mouseenter')
  await trigger(root.querySelector('.var-menu-v2__menu'), 'mouseleave')
  await delay(300)
  expect(root.innerHTML).toMatchSnapshot()

  mockRestore()
})
