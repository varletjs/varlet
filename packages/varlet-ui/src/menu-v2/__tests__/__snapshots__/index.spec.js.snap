// Vitest Snapshot v1

exports[`test menu v2 click trigger 1`] = `"<div data-v-app=\\"\\"><div class=\\"var-menu-v2\\"><!--teleport start--><!--teleport end--></div></div><div style=\\"z-index: 2001; position: absolute; left: 0px; top: 0px; margin: 0px; transform: translate(0px, 0px);\\" class=\\"var-menu-v2__menu var-menu-v2--menu-background-color var-elevation--3\\" data-popper-placement=\\"bottom-start\\"></div>"`;

exports[`test menu v2 click trigger 2`] = `"<div data-v-app=\\"\\"><div class=\\"var-menu-v2\\"><!--teleport start--><!--teleport end--></div></div><div style=\\"z-index: 2001; position: absolute; left: 0px; top: 0px; margin: 0px; transform: translate(0px, 0px); display: none;\\" class=\\"var-menu-v2__menu var-menu-v2--menu-background-color var-elevation--3\\" data-popper-placement=\\"bottom-start\\"></div>"`;

exports[`test menu v2 default style 1`] = `"<transition-stub><div style=\\"z-index: 2002; display: none; position: absolute; left: 0px; top: 0px; margin: 0px; transform: translate(0px, 0px);\\" class=\\"var-menu-v2__menu var-elevation--3\\" data-popper-placement=\\"bottom-start\\"></div></transition-stub>"`;

exports[`test menu v2 hover the menu list 1`] = `"<div style=\\"z-index: 2003; position: absolute; left: 0px; top: 0px; margin: 0px; transform: translate(0px, 0px); display: none;\\" class=\\"var-menu-v2__menu var-menu-v2--menu-background-color var-elevation--3\\" data-popper-placement=\\"bottom-start\\"></div>"`;

exports[`test menu v2 hover trigger and events 1`] = `"<div style=\\"z-index: 2002; position: absolute; left: 0px; top: 0px; margin: 0px; transform: translate(0px, 0px);\\" class=\\"var-menu-v2__menu var-menu-v2--menu-background-color var-elevation--3\\" data-popper-placement=\\"bottom-start\\"></div>"`;

exports[`test menu v2 hover trigger and events 2`] = `"<div style=\\"z-index: 2002; position: absolute; left: 0px; top: 0px; margin: 0px; transform: translate(0px, 0px); display: none;\\" class=\\"var-menu-v2__menu var-menu-v2--menu-background-color var-elevation--3\\" data-popper-placement=\\"bottom-start\\"></div>"`;

exports[`test menu v2 offset 1`] = `"<div style=\\"z-index: 2002; display: none; position: absolute; left: 0px; top: 0px; margin: 0px; transform: translate(100px, 100px);\\" class=\\"var-menu-v2__menu var-menu-v2--menu-background-color var-elevation--3\\" data-popper-placement=\\"bottom-start\\"></div>"`;

exports[`test menu v2 placement 1`] = `"<div style=\\"z-index: 2000; display: none; position: absolute; left: 0px; top: 0px; margin: 0px; bottom: 0px; transform: translate(0px, 0px);\\" class=\\"var-menu-v2__menu var-menu-v2--menu-background-color var-elevation--3\\" data-popper-placement=\\"top\\"></div>"`;

exports[`test menu v2 placement 2`] = `"<div style=\\"z-index: 2000; display: none; position: absolute; left: 0px; top: 0px; margin: 0px; bottom: 0px; transform: translate(0px, 0px);\\" class=\\"var-menu-v2__menu var-menu-v2--menu-background-color var-elevation--3\\" data-popper-placement=\\"top-start\\"></div>"`;

exports[`test menu v2 placement 3`] = `"<div style=\\"z-index: 2000; display: none; position: absolute; left: 0px; top: 0px; margin: 0px; right: 0px; bottom: 0px; transform: translate(0px, 0px);\\" class=\\"var-menu-v2__menu var-menu-v2--menu-background-color var-elevation--3\\" data-popper-placement=\\"top-end\\"></div>"`;

exports[`test menu v2 placement 4`] = `"<div style=\\"z-index: 2000; display: none; position: absolute; left: 0px; top: 0px; margin: 0px; transform: translate(0px, 0px);\\" class=\\"var-menu-v2__menu var-menu-v2--menu-background-color var-elevation--3\\" data-popper-placement=\\"bottom\\"></div>"`;

exports[`test menu v2 placement 5`] = `"<div style=\\"z-index: 2000; display: none; position: absolute; left: 0px; top: 0px; margin: 0px; transform: translate(0px, 0px);\\" class=\\"var-menu-v2__menu var-menu-v2--menu-background-color var-elevation--3\\" data-popper-placement=\\"bottom-start\\"></div>"`;

exports[`test menu v2 placement 6`] = `"<div style=\\"z-index: 2000; display: none; position: absolute; left: 0px; top: 0px; margin: 0px; right: 0px; transform: translate(0px, 0px);\\" class=\\"var-menu-v2__menu var-menu-v2--menu-background-color var-elevation--3\\" data-popper-placement=\\"bottom-end\\"></div>"`;

exports[`test menu v2 placement 7`] = `"<div style=\\"z-index: 2000; display: none; position: absolute; left: 0px; top: 0px; margin: 0px; transform: translate(0px, 0px);\\" class=\\"var-menu-v2__menu var-menu-v2--menu-background-color var-elevation--3\\" data-popper-placement=\\"right\\"></div>"`;

exports[`test menu v2 placement 8`] = `"<div style=\\"z-index: 2000; display: none; position: absolute; left: 0px; top: 0px; margin: 0px; transform: translate(0px, 0px);\\" class=\\"var-menu-v2__menu var-menu-v2--menu-background-color var-elevation--3\\" data-popper-placement=\\"right-start\\"></div>"`;

exports[`test menu v2 placement 9`] = `"<div style=\\"z-index: 2000; display: none; position: absolute; left: 0px; top: 0px; margin: 0px; bottom: 0px; transform: translate(0px, 0px);\\" class=\\"var-menu-v2__menu var-menu-v2--menu-background-color var-elevation--3\\" data-popper-placement=\\"right-end\\"></div>"`;

exports[`test menu v2 placement 10`] = `"<div style=\\"z-index: 2000; display: none; position: absolute; left: 0px; top: 0px; margin: 0px; right: 0px; transform: translate(0px, 0px);\\" class=\\"var-menu-v2__menu var-menu-v2--menu-background-color var-elevation--3\\" data-popper-placement=\\"left\\"></div>"`;

exports[`test menu v2 placement 11`] = `"<div style=\\"z-index: 2000; display: none; position: absolute; left: 0px; top: 0px; margin: 0px; right: 0px; transform: translate(0px, 0px);\\" class=\\"var-menu-v2__menu var-menu-v2--menu-background-color var-elevation--3\\" data-popper-placement=\\"left-start\\"></div>"`;

exports[`test menu v2 placement 12`] = `"<div style=\\"z-index: 2000; display: none; position: absolute; left: 0px; top: 0px; margin: 0px; right: 0px; bottom: 0px; transform: translate(0px, 0px);\\" class=\\"var-menu-v2__menu var-menu-v2--menu-background-color var-elevation--3\\" data-popper-placement=\\"left-end\\"></div>"`;

exports[`test menu v2 placement 13`] = `"<div style=\\"z-index: 2000; display: none; position: absolute; left: 0px; top: 0px; margin: 0px; transform: translate(0px, 0px);\\" class=\\"var-menu-v2__menu var-menu-v2--menu-background-color var-elevation--3\\" data-popper-placement=\\"bottom\\"></div>"`;

exports[`test menu v2 placement 14`] = `"<div style=\\"z-index: 2000; display: none; position: absolute; left: 0px; top: 0px; margin: 0px; transform: translate(0px, 0px);\\" class=\\"var-menu-v2__menu var-menu-v2--menu-background-color var-elevation--3\\" data-popper-placement=\\"bottom-start\\"></div>"`;

exports[`test menu v2 placement 15`] = `"<div style=\\"z-index: 2000; display: none; position: absolute; left: 0px; top: 0px; margin: 0px; right: 0px; transform: translate(0px, 0px);\\" class=\\"var-menu-v2__menu var-menu-v2--menu-background-color var-elevation--3\\" data-popper-placement=\\"bottom-end\\"></div>"`;

exports[`test menu v2 placement 16`] = `"<div style=\\"z-index: 2000; display: none; position: absolute; left: 0px; top: 0px; margin: 0px; bottom: 0px; transform: translate(0px, 0px);\\" class=\\"var-menu-v2__menu var-menu-v2--menu-background-color var-elevation--3\\" data-popper-placement=\\"top\\"></div>"`;

exports[`test menu v2 placement 17`] = `"<div style=\\"z-index: 2000; display: none; position: absolute; left: 0px; top: 0px; margin: 0px; bottom: 0px; transform: translate(0px, 0px);\\" class=\\"var-menu-v2__menu var-menu-v2--menu-background-color var-elevation--3\\" data-popper-placement=\\"top-start\\"></div>"`;

exports[`test menu v2 placement 18`] = `"<div style=\\"z-index: 2000; display: none; position: absolute; left: 0px; top: 0px; margin: 0px; right: 0px; bottom: 0px; transform: translate(0px, 0px);\\" class=\\"var-menu-v2__menu var-menu-v2--menu-background-color var-elevation--3\\" data-popper-placement=\\"top-end\\"></div>"`;

exports[`test menu v2 placement 19`] = `"<div style=\\"z-index: 2000; display: none; position: absolute; left: 0px; top: 0px; margin: 0px; transform: translate(0px, 0px);\\" class=\\"var-menu-v2__menu var-menu-v2--menu-background-color var-elevation--3\\" data-popper-placement=\\"right\\"></div>"`;

exports[`test menu v2 placement 20`] = `"<div style=\\"z-index: 2000; display: none; position: absolute; left: 0px; top: 0px; margin: 0px; right: 0px; transform: translate(0px, 0px);\\" class=\\"var-menu-v2__menu var-menu-v2--menu-background-color var-elevation--3\\" data-popper-placement=\\"left\\"></div>"`;
